Implmentation of R2C FFT

smu160 opened the pull request:
@calebzulawski Draft for #23

smu160 committed:
Add draft implmentation of R2C FFT

smu160 committed:
Merge branch 'main' into feature/r2c

smu160 committed:
Add skeleton of impl based on packing complex nums

smu160 committed:
Remove old mod

smu160 committed:
Initial implementation of r2c fft

smu160 committed:
Add implementation based on working python version

smu160 committed:
Fix formatting

smu160 committed:
Add working implementation of R2C FFT

smu160 committed:
Implement r2c test macro for f32 and f64

smu160 commented:
Hi @calebzulawski,

We now have a working version of R2C FFT. This version is clearly sub-optimal as I made correctness the main priority. I'll measure and tune the implementation for now. In the interim, I recall you mentioned, inter alia, ideas about the public API and the use of traits.

I'm curious to hear your thoughts. Thank you! 

Best,
Saveliy

calebzulawski commented on src/fft.rs:
```diff
@@ -0,0 +1,165 @@
+}
+
+/// Implementation of Real-Valued FFT
+pub fn r2c_fft_f64(input_re: &[f64]) -> (Vec<f64>, Vec<f64>) {
```
Any chance this could be non-allocating and take &mut outputs?

smu160 replied:
Let me now if the checked in fix will suffice

calebzulawski commented:
Looks good! I think starting with correctness is the way to go.  Regarding traits, I think I will probably submit a PR in the future after this has been merged.

smu160 committed:
Use preallocated output arrays

smu160 committed:
Fix formatting

smu160 committed:
Remove extra memory allocations in R2C FFT

smu160 committed:
Fix formatting

smu160 committed:
Add criterion and benchmark

- Benchmark throughput/time of Real-to-Complex (R2C) versus the
  Complex-to-Complex (C2C) FFT on real input data

smu160 committed:
Add intructions for benchmarking using criterion

smu160 committed:
Add subsection before graph

smu160 committed:
Fix benchmark plot rendering issue

smu160 committed:
Add python bindings for Real FFT

smu160 committed:
Add dependencies to run audio visualizer example

smu160 committed:
Merge branch 'main' of github.com:QuState/PhastFT into feature/r2c

smu160 committed:
Remove superfluous macro export

smu160 committed:
Update test hook to run for all features

smu160 committed:
Update CI tests to test all features

smu160 committed:
Add docs for Real FFT

smu160 committed:
Cleanup docs for R2C FFT

smu160 committed:
Add inverse of R2C FFT (f64 only)

smu160 commented:
> Looks good! I think starting with correctness is the way to go. Regarding traits, I think I will probably submit a PR in the future after this has been merged.

Hi @calebzulawski,

I believe we're finally at a point where we can scrutinize the implementation, make necessary changes, and merge. Please feel free to point out any deficiencies and I can work on fixing them. 

Thank you!!

Shnatsel commented on src/fft.rs:
```diff
@@ -0,0 +1,299 @@
+            // Zy = -0.5j * (Z - Zminconj)
+            z_even
+                .iter()
+                .skip(1)
```
`.skip(1)` may sometimes give worse codegen than plain `iter()` over `slice[1..]` so that's something to potentially look into.

Shnatsel commented on src/fft.rs:
```diff
@@ -0,0 +1,299 @@
+                    z_x_re.push(t);
+                    z_x_im.push(u);
+                    z_y_re.push(w);
+                    z_y_im.push(v);
```
Pushing in a loop isn't great, because it bumps 4 counters on every loop iteration, and probably involves a branch on capacity too.

If we could instead iterate over all of these inputs in a chunked way, with `chunks_exact`, write into intermediate buffers of type `scratch_space = [$precision; 4]` (or any other reasonable size) and then `.extend_from_slice(scratch_space)` on each vector, we might save ourselves half the branches, and unlock some SIMD opportunities too now that we're operating on batches of values.

Or we might invoke LLVM's wrath with non-inlined calls to `memcpy()` for anything over `[f32; 4]` or `[f64; 2]` - dunno. Gotta try it and see.

Shnatsel commented on src/fft.rs:
```diff
@@ -0,0 +1,299 @@
+                let zy_re = z_y_re[i];
+                let zy_im = z_y_im[i];
+                let w_re = twiddle_re[i];
+                let w_im = twiddle_im[i];
```
Indexing in a potentially hot loop need to be inspected for bounds checks in godbolt or similar, and a comment added that it's inspected and doesn't have them, or that a `.get_unchecked()` prototype didn't gain you any performance.

If bounds checks do get in the way, try [the usual techniques](https://shnatsel.medium.com/how-to-avoid-bounds-checks-in-rust-without-unsafe-f65e618b4c1e) or amortize them with processing batches of 4 or 8 elements at a time, performing the bounds checks once per slice.

smu160 committed:
Add throughput benchmark for `realfft`

smu160 committed:
Use iterators in untangle step

smu160 committed:
Pre-allocate vectors in lieu of `with_capacity`

smu160 committed:
Replace `skip(1)` by using slices

smu160 committed:
Fuse separate iterators together

smu160 committed:
Use SIMD twiddle factor generator for N/2 >= 16

smu160 committed:
Generate twiddles once for untangling & Planner

smu160 committed:
Merge branch 'main' into feature/r2c

smu160 committed:
Merge branch 'main' into feature/r2c

smu160 committed:
Merge branch 'main' into feature/r2c

smu160 committed:
Fix bug in benchmarks and fix clippy complaints

smu160 committed:
Merge branch 'feature/r2c' of github.com:QuState/PhastFT into feature/r2c

smu160 committed:
Comment out uneeded example code

smu160 committed:
Fix bug in r2c fft using extra planner

- This is a quick, but hacky and inefficient way to use the twiddle
  factors generated by the `Planner`, but it works.

- Add more details to assert statements for debugging

smu160 committed:
Use `Twiddles` iter for untangling step in R2C FFT

smu160 committed:
Move planners out of r2c vs c2c benchmarks

smu160 committed:
Replace naive deinterleave in R2C FFT

smu160 committed:
Add more comments to explain r2c fft

smu160 commented:
Closing in favor of #105
